numba/extending.py

```diff
@@ -0,0 +1,117 @@
+
+import inspect
+
+from numba import types
+
+# Exported symbols
+from .typing.typeof import typeof_impl
+from .targets.imputils import builtin, builtin_cast, implement, impl_attribute
+from .datamodel import models, register_default as register_model
+from .pythonapi import box, unbox, reflect, NativeValue
+
+
+def type_callable(func):
+    """
+    Decorate a function as implementing typing for the callable *func*.
+    *func* can be a callable object (probably a global) or a string
+    denoting a built-in operation (such 'getitem' or '__array_wrap__')
+    """
+    from .typing.templates import CallableTemplate, builtin, builtin_global
+    if not callable(func) and not isinstance(func, str):
+        raise TypeError("`func` should be a function or string")
+    try:
+        func_name = func.__name__
+    except AttributeError:
+        func_name = str(func)
+
+    def decorate(typing_func):
+        def generic(self):
+            return typing_func(self.context)
+
+        name = "%s_CallableTemplate" % (func_name,)
+        bases = (CallableTemplate,)
+        class_dict = dict(key=func, generic=generic)
+        template = type(name, bases, class_dict)
+        builtin(template)
+        if hasattr(func, '__module__'):
+            builtin_global(func, types.Function(template))
+
+    return decorate
+
+
+def overlay(func):
```

Comment:
If the intention of this function is to `overload` and this is the public API perhaps this function should be named as such to avoid confusion? It is noted also that `overload` is ridiculously overloaded as a word and concept already, so there is an argument over what creates less confusion, yet another overload of overload or a rename and users not knowing/realising what `overlay` is.